src/Numerics/DGMethods/DGModel.jl

```diff
@@ -1043,3 +875,646 @@ function continuous_field_gradient!(
     )
     wait(device, event)
 end
+
+function hyperdiff_indexmap(balance_law, ::Type{FT}) where {FT}
+    ns_hyperdiff = number_states(balance_law, Hyperdiffusive())
+    if ns_hyperdiff > 0
+        return varsindices(
+            vars_state(balance_law, Gradient(), FT),
+            fieldnames(vars_state(balance_law, GradientLaplacian(), FT)),
+        )
+    else
+        return nothing
+    end
+end
+
+"""
+    launch_volume_gradients!(dg, state_prognostic, t; dependencies)
+
+Launches horizontal and vertical kernels for computing the volume gradients.
+"""
+function launch_volume_gradients!(dg, state_prognostic, t; dependencies)
+    FT = eltype(state_prognostic)
+    Qhypervisc_grad, _ = dg.states_higher_order
+
+    info = basic_launch_info(dg)
+    workgroup = (info.Nq, info.Nq)
+    ndrange = (info.Nq * info.nrealelem, info.Nq)
+    comp_stream = dependencies
+
+    # If the model direction is EveryDirection, we need to perform
+    # both horizontal AND vertical kernel calls; otherwise, we only
+    # call the kernel corresponding to the model direction `dg.diffusion_direction`
+    if dg.diffusion_direction isa EveryDirection ||
+       dg.diffusion_direction isa HorizontalDirection
+        comp_stream = volume_gradients!(info.device, workgroup)(
+            dg.balance_law,
+            Val(info.dim),
+            Val(info.N),
+            HorizontalDirection(),
+            state_prognostic.data,
+            dg.state_gradient_flux.data,
+            Qhypervisc_grad.data,
+            dg.state_auxiliary.data,
+            dg.grid.vgeo,
+            t,
+            dg.grid.D,
+            Val(hyperdiff_indexmap(dg.balance_law, FT)),
+            dg.grid.topology.realelems,
+            ndrange = ndrange,
+            dependencies = comp_stream,
+        )
+    end
+
+    # Now we call the kernel corresponding to the vertical direction
+    if dg.diffusion_direction isa EveryDirection ||
+       dg.diffusion_direction isa VerticalDirection
+        comp_stream = volume_gradients!(info.device, workgroup)(
+            dg.balance_law,
+            Val(info.dim),
+            Val(info.N),
+            VerticalDirection(),
+            state_prognostic.data,
+            dg.state_gradient_flux.data,
+            Qhypervisc_grad.data,
+            dg.state_auxiliary.data,
+            dg.grid.vgeo,
+            t,
+            dg.grid.D,
+            Val(hyperdiff_indexmap(dg.balance_law, FT)),
+            dg.grid.topology.realelems,
+            # If we are computing the volume gradient in every direction, we
+            # need to increment into the appropriate fields _after_ the
+            # horizontal computation.
+            !(dg.diffusion_direction isa VerticalDirection),
+            ndrange = ndrange,
+            dependencies = comp_stream,
+        )
+    end
```

Comment:
Pardon me if this is a really silly question, due to the fact that I'm still learning Julia and its magical multiple-dispatch features, but could this block of duplicated code be simplified by checking the direction first, assigning to a variable, and then passing only that one variable, so that the function call is the same in both cases? Something like:
```suggestion
    # If the model direction is EveryDirection, we need to perform
    # both horizontal AND vertical kernel calls; otherwise, we only
    # call the kernel corresponding to the model direction `dg.diffusion_direction`
    if dg.diffusion_direction isa EveryDirection ||
       dg.diffusion_direction isa HorizontalDirection
         Direction() = HorizontalDirection()
    elseif dg.diffusion_direction isa EveryDirection ||
       dg.diffusion_direction isa VerticalDirection
         Direction() = VerticalDirection()
     end
      
    # Now we call the kernel with the desired direction
     comp_stream = volume_gradients!(info.device, workgroup)(
         dg.balance_law,
         Val(info.dim),
         Val(info.N),
         Direction(),
         state_prognostic.data,
         dg.state_gradient_flux.data,
         Qhypervisc_grad.data,
         dg.state_auxiliary.data,
         dg.grid.vgeo,
         t,
         dg.grid.D,
         Val(hyperdiff_indexmap(dg.balance_law, FT)),
         dg.grid.topology.realelems,
         # If we are computing the volume gradient in every direction, we
         # need to increment into the appropriate fields _after_ the
         # horizontal computation.
         !(dg.diffusion_direction isa VerticalDirection),
         ndrange = ndrange,
         dependencies = comp_stream,
     )
```
I believe this would reduce duplication of code, hence it would help with code maintenance